File: src/main/scala/llmTests/a201706/scala/gpt4/SequenceHelpersImpl.scala
package llmTests.a201706.scala.gpt4

import llmTests.a201706.scala.Sequence
import llmTests.a201706.scala.SequenceHelpers
import llmTests.shared.scala.Pair
import java.util
import java.util.function.BinaryOperator

class SequenceHelpersImpl extends SequenceHelpers {
  override def of[X](x: X): Sequence[X] = new Sequence[X] {
    override def nextElement: X = x
  }

  override def cyclic[X](l: util.List[X]): Sequence[X] = new Sequence[X] {
    private var i = 0

    override def nextElement: X = {
      val elem = l.get(i)
      i = if (i < l.size() - 1) i + 1 else 0
      elem
    }
  }

  override def incrementing(start: Int, increment: Int): Sequence[Integer] =
    new Sequence[Integer] {
      private var currVal = start
      override def nextElement: Integer = {
        val tmp = Integer.valueOf(currVal)
        currVal += increment
        tmp
      }
    }

  override def accumulating[X](input: Sequence[X], op: BinaryOperator[X]):
  Sequence[X] = new Sequence[X] {
    private var acc = input.nextElement

    override def nextElement: X = {
      val tmp = acc
      acc = op.apply(acc, input.nextElement)
      tmp
    }
  }

  override def zip[X](input: Sequence[X]): Sequence[Pair[X, Integer]] =
    new Sequence[Pair[X, Integer]] {
      private var idx = 0
      override def nextElement: Pair[X, Integer] = {
        val tmp = new Pair[X, Integer](input.nextElement, Integer.valueOf(idx))
        idx += 1
        tmp
      }
    }
}

File: src/main/scala/llmTests/a201706/java/gpt4/SequenceHelpersImpl.java
package llmTests.a201706.java.gpt4;

import java.util.Iterator;
import java.util.List;
import java.util.function.BinaryOperator;

import llmTests.a201706.java.Sequence;
import llmTests.a201706.java.SequenceHelpers;
import llmTests.shared.java.Pair;

public class SequenceHelpersImpl implements SequenceHelpers {
    public <X> Sequence<X> of(X x) {
        return new Sequence<X>(){
            public X nextElement() {
                return x;
            }
        };
    }

    public <X> Sequence<X> cyclic(List<X> l) {
        return new Sequence<X>(){
            private Iterator<X> it = l.iterator();
            public X nextElement() {
                if (!it.hasNext()) it = l.iterator();
                return it.next();
            }
        };
    }

    public Sequence<Integer> incrementing(int start, int increment) {
        return new Sequence<Integer>(){
            private int current = start;
            public Integer nextElement() {
                int old = current;
                current += increment;
                return old;
            }
        };
    }


    public <X> Sequence<Pair<X, Integer>> zip(Sequence<X> input) {
        return new Sequence<Pair<X, Integer>>() {

            private int index = 0;

            @Override
            public Pair<X, Integer> nextElement() {
                X elem = input.nextElement();
                int currentIndex = index++;
                return new Pair<X, Integer>(elem, currentIndex);
            }
        };
    }

    public <X> Sequence<X> accumulating(Sequence<X> input, BinaryOperator<X> op) {
        return new Sequence<X>(){
            private X accumulated = input.nextElement();
            public X nextElement() {
                X old = accumulated;
                accumulated = op.apply(accumulated, input.nextElement());
                return old;
            }
        };
    }
}